File: code/mathematical-algorithms/2Sum/2sum.go
// Part of Cosmos by OpenGenus Foundation
package main

import "fmt"

func twoSum(data []int, index map[int]bool, target int) {
	for i := range data {
		if index[target-data[i]] {
			fmt.Printf("We can get %d use %d and %d in %v\n", target, data[i], target-data[i], data)
			return
		}
	}
	fmt.Printf("We can't get %d by combime two numbers in %v\n", target, data)
	return
}

func main() {
	data := []int{1, 2, 3, 4, 5, 6, 7, 7, 8, 9, 10, 12, 13}
	index := make(map[int]bool)

	for i := range data {
		index[data[i]] = true
	}

	twoSum(data, index, 5)
	twoSum(data, index, 22)
	twoSum(data, index, 33)
	twoSum(data, index, 123)
}

File: code/mathematical-algorithms/2Sum/2sum.rb
# Part of Cosmos by OpenGenus Foundation
# Given an array of integers, return the indices of the two numbers such that they add up to a specific target.

def two_sum(list, target)
  buffer = {}
  list.each_with_index do |val, idx|
    return [buffer[target - val], idx] if buffer.key?(target - val)
    buffer[val] = idx
  end
  [nil, nil]
end

p two_sum([3, 5, 7, 0, -3, -2, -3], 4)
p two_sum([3, 5, 0, -3, -2, -3], 4)
